### PR TITLE
[SPARK-32154][SQL] Use ExpressionEncoder for the return type of ScalaUDF to convert to catalyst type

### PR DESCRIPTION
**Pull request by Ngone51**
### What changes were proposed in this pull request?

This PR proposes to use `ExpressionEncoder` for the return type of ScalaUDF to convert to the catalyst type, instead of using `CatalystTypeConverters`. 

Note, this change only takes effect for typed Scala UDF since its the only case where we know the type tag of the raw type.

### Why are the changes needed?

Users now could register a UDF with `Instant`/`LocalDate` as return types even with `spark.sql.datetime.java8API.enabled=false`. However, the UDF can not really be used.
For example, if we try:

```scala
scala> sql("set spark.sql.datetime.java8API.enabled=false")
scala> spark.udf.register("buildDate", udf{ d: String => java.time.LocalDate.parse(d) })
scala> Seq("2020-07-02").toDF("d").selectExpr("CAST(buildDate(d) AS STRING)").show
```
Then, we will hit the error:
```scala
java.lang.ClassCastException: java.time.LocalDate cannot be cast to java.sql.Date
  at org.apache.spark.sql.catalyst.CatalystTypeConverters$DateConverter$.toCatalystImpl(CatalystTypeConverters.scala:304)
  at org.apache.spark.sql.catalyst.CatalystTypeConverters$CatalystTypeConverter.toCatalyst(CatalystTypeConverters.scala:107)
  at org.apache.spark.sql.catalyst.CatalystTypeConverters$.$anonfun$createToCatalystConverter$2(CatalystTypeConverters.scala:425)
  at org.apache.spark.sql.catalyst.expressions.ScalaUDF.eval(ScalaUDF.scala:1169)
...
```
as it actually requires enabling `spark.sql.datetime.java8API.enabled` when using the UDF. And I think this could make users get confused.

This happens because when registering the UDF,  Spark actually uses `ExpressionEncoder` to ser/deser types. However, when using UDF, Spark uses `CatalystTypeConverters`, which is under control of `spark.sql.datetime.java8API.enabled`, to ser/deser types. Therefore, Spark would fail to convert the Java8 date time types.

If we could also use `ExpressionEncoder` to ser/deser types for the return type, similar to what we do for the input parameter types, then, UDF could support Instant/LocalDate, event other combined complex types as well.

### Does this PR introduce _any_ user-facing change?

Yes. Before this PR, if users run the demo above, they would hit the error. After this PR, the demo will run successfully.

### How was this patch tested?

Updated 2 tests and added a new one for combined types of `Instant` and `LocalDate`.

### CODE REVIEW

**Comment by Ngone51** [comment]
ping @maropu @cloud-fan @MaxGekk please take a look, thanks!

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/encoders/ExpressionEncoder.scala [hunk 1]
```diff
@@ -285,7 +285,7 @@ case class ExpressionEncoder[T](
       // transform attributes accessors.
       objDeserializer.transform {
         case UnresolvedExtractValue(GetColumnByOrdinal(0, _),
-            Literal(part: UTF8String, StringType)) =>
+        Literal(part: UTF8String, StringType)) =>
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/encoders/ExpressionEncoder.scala:288]: unnecessary change

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/ScalaUDF.scala [hunk 2]
```diff
@@ -102,6 +105,28 @@ case class ScalaUDF(
...
 
+  /**
+   * Create the converter which converts the scala data type to the catalyst data type for
+   * the return data type of udf function. We'd use `ExpressionEncoder` to create the
+   * converter for typed ScalaUDF only, since its the only case where we know the  type tag
+   * of the return data type of udf function.
+   * @param dataType return type of function
+   * @return the catalyst converter
+   */
+  private def catalystConverter(dataType: DataType): Any => Any = {
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/ScalaUDF.scala:116]: we don't need this parameter. `dataType` is defined in the constructor parameters.

File: sql/core/src/test/scala/org/apache/spark/sql/UDFSuite.scala [hunk 3]
```diff
@@ -504,23 +508,34 @@ class UDFSuite extends QueryTest with SharedSparkSession {
...
-      assert(df.collect().toSeq === Seq(Row(expected)))
-    }
+    val expected = java.time.LocalDate.parse("2019-02-27").toString
+    val plusDay = udf((i: java.time.LocalDate) => i.plusDays(1))
+    val df = spark.sql("SELECT DATE '2019-02-26' as d")
+      .select(plusDay('d).cast(StringType))
+    assert(df.collect().toSeq === Seq(Row(expected)))
+  }
+
+  test("Using combined types of Instant/LocalDate in UDF") {
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/UDFSuite.scala:530]: can we add one more test to combine `Instant` and `Timestamp`? This is not possible before this PR.

**Review by viirya — COMMENTED** [review]
So based on the description, after this change, UDF is not controlled by `spark.sql.datetime.java8API.enabled`?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/encoders/ExpressionEncoder.scala [hunk 4]
```diff
@@ -305,6 +305,14 @@ case class ExpressionEncoder[T](
     StructField(s.name, s.dataType, s.nullable)
   })
 
+  def dataTypeAndNullable(): (DataType, Boolean) = {
+    if (isSerializedAsStruct) {
+      (schema, true)
```
Comment (viirya) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/encoders/ExpressionEncoder.scala:310]: nullable should be `objSerializer.nullable`?

**Comment by cloud-fan** [comment]
> UDF is not controlled by spark.sql.datetime.java8API.enabled

UDF shouldn't be. The config is only applicable if users want to get data from Spark, and this config decides the datetime class to use. But UDF returns data to Spark, and it's better if Spark can accept all the datetime classes.

**Comment by viirya** [comment]
I see. Make sense to me.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/ScalaUDF.scala [hunk 5]
```diff
@@ -102,6 +105,28 @@ case class ScalaUDF(
...
+    if (returnEncoder.isDefined) {
```
Comment (maropu) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/ScalaUDF.scala:117]: nit format: how about using `.map` instead of `isDefined`?
```
    returnEncoder.map { enc =>
      val toRow = enc.createSerializer()
      if (enc.isSerializedAsStruct) {
```

File: sql/core/src/test/scala/org/apache/spark/sql/UDFSuite.scala [hunk 6]
```diff
@@ -504,23 +508,34 @@ class UDFSuite extends QueryTest with SharedSparkSession {
...
+    val ts = "2019-02-26T23:59:59Z"
+    val date = "2019-02-26"
+    val expectedDate = sql(s"SELECT CAST(DATE '$date' AS STRING)").collect().head.getString(0)
+    val expectedIns = sql(s"SELECT CAST(TIMESTAMP '$ts' AS STRING)").collect().head.getString(0)
+    spark.udf.register("toDateTime", udf((d: LocalDate, i: Instant) => DateTimeResult(d, i)))
```
Comment (maropu) [inline sql/core/src/test/scala/org/apache/spark/sql/UDFSuite.scala:535]: If `udf((d: LocalDate, i: Instant) => null))`, it seems `ExpressionEncoder` will throw an exception?

Comment (maropu) [reply sql/core/src/test/scala/org/apache/spark/sql/UDFSuite.scala:535]: I think we need to test the case: `udf((d: LocalDate, i: Instant) => DateTimeResult(null, null))`.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/ScalaUDF.scala (hunk 2 above)
Comment (Ngone51) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/ScalaUDF.scala]: ok

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala [hunk 7]
```diff
@@ -2819,13 +2819,12 @@ class Analyzer(
 
       case p => p transformExpressionsUp {
 
-        case udf @ ScalaUDF(_, _, inputs, _, _, _, _)
-            if udf.inputPrimitives.contains(true) =>
+        case udf: ScalaUDF if udf.inputPrimitives.contains(true) =>
```
[the remark this answers is not in the document]
Comment (Ngone51) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:2822]: yea, sounds good.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/ScalaUDF.scala [hunk 8]
```diff
@@ -36,6 +36,8 @@ import org.apache.spark.sql.types.{AbstractDataType, AnyDataType, DataType, User
  * @param inputEncoders ExpressionEncoder for each input parameters. For a input parameter which
  *                      serialized as struct will use encoder instead of CatalystTypeConverters to
  *                      convert internal value to Scala value.
+ * @param returnEncoder ExpressionEncoder for the return type of function. It's only defined when
```
[the remark this answers is not in the document]
Comment (Ngone51) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/ScalaUDF.scala]: yea, it's more consistent compares to `inputEncoders `

File: sql/core/src/test/scala/org/apache/spark/sql/UDFSuite.scala (hunk 6 above)
Comment (Ngone51) [reply sql/core/src/test/scala/org/apache/spark/sql/UDFSuite.scala]: > If udf((d: LocalDate, i: Instant) => null)), it seems ExpressionEncoder will throw an exception?

It will since Spark does not support top-level null values.


> I think we need to test the case: udf((d: LocalDate, i: Instant) => DateTimeResult(null, null)).

Improved test for this case.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/ScalaUDF.scala [hunk 9]
```diff
@@ -102,6 +105,28 @@ case class ScalaUDF(
...
   }
 
+  /**
+   * Create the converter which converts the scala data type to the catalyst data type for
+   * the return data type of udf function. We'd use `ExpressionEncoder` to create the
+   * converter for typed ScalaUDF only, since its the only case where we know the  type tag
+   * of the return data type of udf function.
+   * @param dataType return type of function
+   * @return the catalyst converter
+   */
```
[the remark this answers is not in the document]
Comment (Ngone51) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/ScalaUDF.scala:113]: I'm a little bit hesitant about this. I named it as `catalystConverter` because we've had `scalaConverter`. It's more consistent naming in this way. And the reason we named `scalaConverter` instead of `createToScalaConverter` is that it cause the name collision with `CatalystTypeConverters.createToScalaConverter`.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/encoders/ExpressionEncoder.scala [hunk 10]
```diff
@@ -285,7 +285,7 @@ case class ExpressionEncoder[T](
       // transform attributes accessors.
       objDeserializer.transform {
         case UnresolvedExtractValue(GetColumnByOrdinal(0, _),
-            Literal(part: UTF8String, StringType)) =>
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/encoders/ExpressionEncoder.scala:288]: the previous indentation is correct.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/ScalaUDF.scala [hunk 11]
```diff
@@ -62,7 +65,7 @@ case class ScalaUDF(
    * Nil(has same effect with all false) and analyzer will skip null-handling
    * on them.
    */
-  def inputPrimitives: Seq[Boolean] = {
+  val inputPrimitives: Seq[Boolean] = {
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/ScalaUDF.scala:68]: lazy val?

Comment (maropu) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/ScalaUDF.scala:68]: +1

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/ScalaUDF.scala (hunk 9 above)
[the remark this answers is not in the document]
Comment (maropu) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/ScalaUDF.scala:113]: Looks okay as it is.

File: sql/core/src/test/scala/org/apache/spark/sql/UDFSuite.scala (hunk 6 above)
Comment (maropu) [reply sql/core/src/test/scala/org/apache/spark/sql/UDFSuite.scala]: The previous behaviour looks like a bug? cc: @cloud-fan 
```
// v3.0
scala> val dtf = java.time.format.DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm:ss")
scala> val timestamp = java.sql.Timestamp.valueOf("2019-02-26 23:59:59")
scala> val df = Seq((timestamp)).toDF("t")
scala> spark.udf.register("scalaUdf", udf((t: java.sql.Timestamp) => null))
scala> df.selectExpr("scalaUdf(t)").show()
+------+
|UDF(t)|
+------+
|  null|
+------+

// This PR
scala> val dtf = java.time.format.DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm:ss")
scala> val timestamp = java.sql.Timestamp.valueOf("2019-02-26 23:59:59")
scala> val df = Seq((timestamp)).toDF("t")
scala> spark.udf.register("scalaUdf", udf((t: java.sql.Timestamp) => null))
java.lang.ClassNotFoundException: scala.Null
  at scala.reflect.internal.util.AbstractFileClassLoader.findClass(AbstractFileClassLoader.scala:72)
  at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:588)
  at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:521)
  at java.base/java.lang.Class.forName0(Native Method)
  at java.base/java.lang.Class.forName(Class.java:398)
  at scala.reflect.runtime.JavaMirrors$JavaMirror.javaClass(JavaMirrors.scala:589)
  at scala.reflect.runtime.JavaMirrors$JavaMirror.$anonfun$classToJava$1(JavaMirrors.scala:1265)
  at scala.reflect.runtime.TwoWayCaches$TwoWayCache.$anonfun$toJava$1(TwoWayCaches.scala:61)
  at scala.reflect.runtime.TwoWayCaches$TwoWayCache.toJava(TwoWayCaches.scala:57)
  at scala.reflect.runtime.JavaMirrors$JavaMirror.classToJava(JavaMirrors.scala:1257)
  at scala.reflect.runtime.JavaMirrors$JavaMirror.typeToJavaClass(JavaMirrors.scala:1351)
  at scala.reflect.runtime.JavaMirrors$JavaMirror.runtimeClass(JavaMirrors.scala:227)
  at scala.reflect.runtime.JavaMirrors$JavaMirror.runtimeClass(JavaMirrors.scala:68)
  at org.apache.spark.sql.catalyst.encoders.ExpressionEncoder$.apply(ExpressionEncoder.scala:56)
  at org.apache.spark.sql.functions$.udf(functions.scala:4542)
  ... 47 elided
```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/ScalaUDF.scala (hunk 9 above)
[the remark this answers is not in the document]
Comment (Ngone51) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/ScalaUDF.scala:113]: Thanks.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/encoders/ExpressionEncoder.scala (hunk 10 above)
Comment (Ngone51) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/encoders/ExpressionEncoder.scala:288]: Oh, I see. (Actually, I never touch it previously. It seems like another Github mistake.)

File: sql/core/src/test/scala/org/apache/spark/sql/UDFSuite.scala (hunk 6 above)
Comment (Ngone51) [reply sql/core/src/test/scala/org/apache/spark/sql/UDFSuite.scala]: I see...Spark 3.0 use `ScalaReflection.schemaFor[RT]` to get the data type for the return type and it supports `null`. And in this PR, we use `ExpressionEncoder[RT]()`, which doesn't support `null`.

Comment (Ngone51) [reply sql/core/src/test/scala/org/apache/spark/sql/UDFSuite.scala]: > > If udf((d: LocalDate, i: Instant) => null)), it seems ExpressionEncoder will throw an exception?
> 
> It will since Spark does not support top-level null values.


After taking another look, I think we should at least be able to support `udf((d: LocalDate, i: Instant) => null.asInstantOf[SomeType])`. In this case, the output of the UDF is a value of column instead of the top-level row value. Previously, ScalaUDF supports it as well.

**Comment by Ngone51** [comment]
retest this please.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/ScalaUDF.scala [hunk 12]
```diff
@@ -102,6 +105,25 @@ case class ScalaUDF(
...
   }
 
+  /**
+   * Create the converter which converts the scala data type to the catalyst data type for
+   * the return data type of udf function. We'd use `ExpressionEncoder` to create the
+   * converter for typed ScalaUDF only, since its the only case where we know the type tag
+   * of the return data type of udf function.
+   */
+  private def catalystConverter: Any => Any = outputEncoder.map { enc =>
+    val toRow = enc.createSerializer()
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/ScalaUDF.scala:115]: nit: `val toRow = enc.createSerializer().asInstanceOf[Any => Any]`

File: sql/core/src/test/scala/org/apache/spark/sql/UDFSuite.scala [hunk 13]
```diff
@@ -504,23 +510,87 @@ class UDFSuite extends QueryTest with SharedSparkSession {
...
+      udf((d: LocalDate, i: Instant) => LocalDateInstantType(d, i)))
+    checkAnswer(df.selectExpr(s"buildLocalDateInstantType(d, i) as di")
+      .select('di.cast(StringType)),
+      Row(s"[$expectedDate, $expectedInstant]") :: Nil)
+
+    // test null cases
+    spark.udf.register("buildLocalDateInstantType",
+      udf((d: LocalDate, i: Instant) => LocalDateInstantType(null, null)))
+    checkAnswer(df.selectExpr("buildLocalDateInstantType(d, i) as di")
+      .select('di.cast(StringType)),
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/UDFSuite.scala:554]: we don't need cast here, as the result is `Row(null, null)`

File: sql/core/src/test/scala/org/apache/spark/sql/UDFSuite.scala [hunk 14]
```diff
@@ -504,23 +510,87 @@ class UDFSuite extends QueryTest with SharedSparkSession {
...
+      Row("[,]"))
+
+    spark.udf.register("buildLocalDateInstantType",
+      udf((d: LocalDate, i: Instant) => null.asInstanceOf[LocalDateInstantType]))
+    checkAnswer(df.selectExpr("buildLocalDateInstantType(d, i) as di")
+      .select('di.cast(StringType)),
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/UDFSuite.scala:560]: ditto

File: sql/core/src/test/scala/org/apache/spark/sql/UDFSuite.scala [hunk 15]
```diff
@@ -504,23 +510,87 @@ class UDFSuite extends QueryTest with SharedSparkSession {
...
+      udf((t: Timestamp, i: Instant) => TimestampInstantType(t, i)))
+    checkAnswer(df.selectExpr("buildTimestampInstantType(t, i) as ti")
+      .select('ti.cast(StringType)),
+      Row(s"[$expectedTimestamp, $expectedInstant]"))
+
+    // test null cases
+    spark.udf.register("buildTimestampInstantType",
+      udf((t: Timestamp, i: Instant) => TimestampInstantType(null, null)))
+    checkAnswer(df.selectExpr("buildTimestampInstantType(t, i) as ti")
+      .select('ti.cast(StringType)),
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/UDFSuite.scala:586]: ditto

File: sql/core/src/test/scala/org/apache/spark/sql/UDFSuite.scala [hunk 16]
```diff
@@ -504,23 +510,87 @@ class UDFSuite extends QueryTest with SharedSparkSession {
...
+      Row("[,]"))
+
+    spark.udf.register("buildTimestampInstantType",
+      udf((t: Timestamp, i: Instant) => null.asInstanceOf[TimestampInstantType]))
+    checkAnswer(df.selectExpr("buildTimestampInstantType(t, i) as ti")
+      .select('ti.cast(StringType)),
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/UDFSuite.scala:592]: ditto

**Review by cloud-fan — APPROVED** [review]

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/encoders/ExpressionEncoder.scala [hunk 17]
```diff
@@ -305,6 +305,14 @@ case class ExpressionEncoder[T](
     StructField(s.name, s.dataType, s.nullable)
   })
 
+  def dataTypeAndNullable: (DataType, Boolean) = {
+    if (isSerializedAsStruct) {
+      (schema, objSerializer.nullable)
+    } else {
+      (schema.head.dataType, objSerializer.nullable)
+    }
```
Comment (maropu) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/encoders/ExpressionEncoder.scala:313]: nit: 
```
    val dataType = if (isSerializedAsStruct) schema else schema.head.dataType
    (dataType, objSerializer.nullable)
```

File: sql/core/src/main/scala/org/apache/spark/sql/expressions/UserDefinedFunction.scala [hunk 18]
```diff
@@ -94,6 +94,7 @@ private[spark] case class SparkUserDefinedFunction(
     f: AnyRef,
     dataType: DataType,
     inputEncoders: Seq[Option[ExpressionEncoder[_]]] = Nil,
+    returnEncoder: Option[ExpressionEncoder[_]] = None,
```
Comment (maropu) [inline sql/core/src/main/scala/org/apache/spark/sql/expressions/UserDefinedFunction.scala:97]: nit: `returnEncoder` -> `outputEncoder`?

File: sql/core/src/test/scala/org/apache/spark/sql/UDFSuite.scala [hunk 19]
```diff
@@ -504,23 +510,83 @@ class UDFSuite extends QueryTest with SharedSparkSession {
...
+    // test null cases
+    spark.udf.register("buildTimestampInstantType",
+      udf((t: Timestamp, i: Instant) => TimestampInstantType(null, null)))
+    checkAnswer(df.selectExpr("buildTimestampInstantType(t, i) as ti"),
+      Row(Row(null, null)))
+
+    spark.udf.register("buildTimestampInstantType",
+      udf((t: Timestamp, i: Instant) => null.asInstanceOf[TimestampInstantType]))
+    checkAnswer(df.selectExpr("buildTimestampInstantType(t, i) as ti"),
+      Row(null))
```
Comment (maropu) [inline sql/core/src/test/scala/org/apache/spark/sql/UDFSuite.scala:589]: Thanks for adding this test. To check the unsupported case, could you add tests for the toplevel-null case (catch encoder exceptions)? https://github.com/apache/spark/pull/28979#discussion_r450548418

**Comment by maropu** [comment]
LGTM except for the minor comments.

File: sql/core/src/test/scala/org/apache/spark/sql/UDFSuite.scala (hunk 19 above)
Comment (Ngone51) [reply sql/core/src/test/scala/org/apache/spark/sql/UDFSuite.scala:589]: added in 8314a04

**Comment by Ngone51** [comment]
retest this please

**Comment by maropu** [comment]
retest this please

File: sql/core/src/main/scala/org/apache/spark/sql/functions.scala [hunk 20]
```diff
@@ -4521,9 +4522,10 @@ object functions {
    * @since 1.3.0
    */
   def udf[RT: TypeTag](f: Function0[RT]): UserDefinedFunction = {
-    val ScalaReflection.Schema(dataType, nullable) = ScalaReflection.schemaFor[RT]
+    val outputEncoder = Try(ExpressionEncoder[RT]()).toOption
+    val ScalaReflection.Schema(dataType, nullable) = outputEncoder.map(_.dataTypeAndNullable).getOrElse(ScalaReflection.schemaFor[RT])
```
Comment (Ngone51) [inline sql/core/src/main/scala/org/apache/spark/sql/functions.scala:4526]: Hi all, in the latest update, acquring schema can fallback to previous `ScalaReflection.schemaFor[RT]` when `ExpressionEncoder[RT]()` fails. In this way, we can support both`udf((i: String) => null)`(return type is `NullType`) and `udf((i: String) => null.asInstanceOf[String])` (return type is `StringType`) at the same time and avoid behaviour change. cc @maropu

Comment (maropu) [reply sql/core/src/main/scala/org/apache/spark/sql/functions.scala:4526]: Ah, I see. Keeping the current behaviour looks better to me. WDYT? @cloud-fan

File: sql/core/src/main/scala/org/apache/spark/sql/UDFRegistration.scala [hunk 21]
```diff
@@ -173,15 +174,11 @@ class UDFRegistration private[sql] (functionRegistry: FunctionRegistry) extends
     }
     */
 
-  /**
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/UDFRegistration.scala:176]: we need to keep the comment.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/ScalaUDF.scala [hunk 22]
```diff
@@ -102,6 +105,25 @@ case class ScalaUDF(
...
+   * the return data type of udf function. We'd use `ExpressionEncoder` to create the
+   * converter for typed ScalaUDF only, since its the only case where we know the type tag
+   * of the return data type of udf function.
+   */
+  private def catalystConverter: Any => Any = outputEncoder.map { enc =>
+    val toRow = enc.createSerializer().asInstanceOf[Any => Any]
+    if (enc.isSerializedAsStruct) {
+      value: Any =>
+        if (value == null) null
+        else toRow(value).asInstanceOf[InternalRow]
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/ScalaUDF.scala:119]: nit: can we put if-else in one line?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/ScalaUDF.scala [hunk 23]
```diff
@@ -102,6 +105,25 @@ case class ScalaUDF(
...
+    } else {
+      value: Any =>
+        if (value == null) null
+        else toRow(value).asInstanceOf[InternalRow].get(0, dataType)
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/ScalaUDF.scala:123]: ditto

File: sql/core/src/main/scala/org/apache/spark/sql/functions.scala (hunk 20 above)
Comment (cloud-fan) [reply sql/core/src/main/scala/org/apache/spark/sql/functions.scala:4526]: +1 to avoid breaking changes.

File: sql/core/src/test/scala/org/apache/spark/sql/UDFSuite.scala [hunk 24]
```diff
@@ -504,23 +511,92 @@ class UDFSuite extends QueryTest with SharedSparkSession {
...
+  }
+
+  test("SPARK-32154: return null with or without explicit type") {
+    // without explicit type
+    assert(udf((i: String) => null)
+      .asInstanceOf[SparkUserDefinedFunction].dataType === NullType)
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/UDFSuite.scala:596]: can we test it with `checkAnswer` as well?

File: sql/core/src/main/scala/org/apache/spark/sql/functions.scala (hunk 20 above)
Comment (Ngone51) [reply sql/core/src/main/scala/org/apache/spark/sql/functions.scala:4526]: I compared the support types between `schemaFor` between `serializerFor`(which is called by `ExpressionEncoder` indeed). It turns out that we can fallback three types: `NullType`, `CalendarInterval`, `Decimal`.

(Actually, I'm thinking why `serializerFor` doesn't support them. )

Comment (maropu) [reply sql/core/src/main/scala/org/apache/spark/sql/functions.scala:4526]: > I compared the support types between schemaFor between serializerFor(which is called by ExpressionEncoder indeed). It turns out that we can fallback three types: NullType, CalendarInterval, Decimal.

Oh, I see. So, this update looks okay to me.

Comment (Ngone51) [reply sql/core/src/main/scala/org/apache/spark/sql/functions.scala:4526]: Oh..wait!!

I just realized that we should actually compare `schemaFor` and `serializerForType`(rather than `serializerFor`).  `serializerForType` will call `dataTypeFor` first.  `dataTypeFor` actually supports `NullType`, `CalendarInterval`, `Decimal` as well.

The reason `ExpressionEncoder` doesn't support `NullType` is, before `ExpressionEncoder` calls `serializerForType`, it calls `mirror.runtimeClass(tpe)` first:

https://github.com/apache/spark/blob/fab4ca5156d5e1cc0e976c7c27b28a12fa61eb6d/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/encoders/ExpressionEncoder.scala#L56-L58

and `null` (without explicit data type) can fail at `mirror.runtimeClass(tpe)`:
```scala
java.lang.ClassNotFoundException: scala.Null
  at scala.reflect.internal.util.AbstractFileClassLoader.findClass(AbstractFileClassLoader.scala:72)
  at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:588)
  at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:521)
  at java.base/java.lang.Class.forName0(Native Method)
  at java.base/java.lang.Class.forName(Class.java:398)
  at scala.reflect.runtime.JavaMirrors$JavaMirror.javaClass(JavaMirrors.scala:589)
  at scala.reflect.runtime.JavaMirrors$JavaMirror.$anonfun$classToJava$1(JavaMirrors.scala:1265)
  at scala.reflect.runtime.TwoWayCaches$TwoWayCache.$anonfun$toJava$1(TwoWayCaches.scala:61)
  at scala.reflect.runtime.TwoWayCaches$TwoWayCache.toJava(TwoWayCaches.scala:57)
  at scala.reflect.runtime.JavaMirrors$JavaMirror.classToJava(JavaMirrors.scala:1257)
  at scala.reflect.runtime.JavaMirrors$JavaMirror.typeToJavaClass(JavaMirrors.scala:1351)
  at scala.reflect.runtime.JavaMirrors$JavaMirror.runtimeClass(JavaMirrors.scala:227)
  at scala.reflect.runtime.JavaMirrors$JavaMirror.runtimeClass(JavaMirrors.scala:68)
  at org.apache.spark.sql.catalyst.encoders.ExpressionEncoder$.apply(ExpressionEncoder.scala:56)
  at org.apache.spark.sql.functions$.udf(functions.scala:4542)
  ... 47 elided
```

File: sql/core/src/test/scala/org/apache/spark/sql/UDFSuite.scala (hunk 24 above)
Comment (Ngone51) [reply sql/core/src/test/scala/org/apache/spark/sql/UDFSuite.scala]: sure

**Review by cloud-fan — APPROVED** [review]

**Review by viirya — APPROVED** [review]

**Comment by viirya** [comment]
retest this please

**Comment by viirya** [comment]
retest this please

File: sql/core/src/main/scala/org/apache/spark/sql/functions.scala (hunk 20 above)
Comment (maropu) [reply sql/core/src/main/scala/org/apache/spark/sql/functions.scala:4526]: Thanks for the explanation! LGTM

**Review by maropu — APPROVED** [review]

**Comment by maropu** [comment]
retest this please

**Comment by maropu** [comment]
retest this please

**Comment by cloud-fan** [comment]
retest this please

**Comment by maropu** [comment]
retest this please

**Comment by maropu** [comment]
retest this please

**Comment by maropu** [comment]
Thanks! Merged to master.

**Comment by Ngone51** [comment]
thanks all!
